### ISSUE TITLE
Bug: no-shadow doesn't special-case `this` type hint for callbacks

### ISSUE THREAD

**Issue by mqudsi** [typescript-eslint/typescript-eslint#7550]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=GYVwdgxgLglg9mABMOcA8AVANIgqjgYUQF5FwATAU2BjEvID4AKAWAChFEoALGAZwBciDAG0AulnacIAQwA2cgEYyIAayFMe-IQRwA3eSEpDsiWlQAeQsCAC2iygCccMx45kBPE%2BICUJBniSHFy8fACCjgDmAPw67D5CuOKIAN5SiAD0GYgAkpFgcI6UZDZ89IgGjjAyinLFAO6uYLSRfOmyCspqJIgdSiqqANzpWuFRPaMRkcPBPI5w9Yh0iwCiboVMAEQAcnBQZrYADnW2lGBQ9Js%2BMwC%2BM0A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUFs%2B5AdgPYC0yAFgIYAmhA7umANri6sSSLTSHSQA0LNrixC2kMoQCWyAC4NIFePEiDREAL6qhAXRab1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
const foo = function <T, U, C = undefined>(
  this: T[],
  callback: (this: C, value: T, index: number, array: T[]) => U, 
  thisArg?: C
): U[] {
  // Ignore unused variable warnings
  callback = callback;
  thisArg = thisArg;
  throw new Error("Not implemented");
};
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  {
  "rules": {
            "no-shadow": [
              "error",
              {
                "hoist": "all"
              }
            ]
  }
}
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I should be able to declare an infinite number of strongly-typed function callbacks, each with its own `this` type hint, without `no-shadow` complaining that `this` has already been declared.

Note that I'm *not* declaring a variable (`this` exists whether I like it or not), I'm just giving it a type.

### Actual Result

typescript-eslint thinks I'm declaring a new variable with the name `this` and sees it to be in conflict with the outer-scoped `this`.
